### PR TITLE
Unify method pairs with and without Type param

### PR DESCRIPTION
**Pull request by aheejin**
As suggested in
https://github.com/WebAssembly/binaryen/pull/6181#discussion_r1427188670, using `std::optional<Type>`, this unifies two different versions of `make***`, for  block-like structures (`block`, `if`, `loop`, `try`, and `try_table`) with and without a type parameter.

This also allows unifying of `finalize` methods, with and without a type. This also sets `breakability` argument of `Block::finalize` to `Unknown` so we can only have one `Block::finalize` that handles all cases.

This also adds an optional `std::optional<Type> type` parameter to `blockifyWithName`, and `makeSequence` functions in `wasm-builder.h`. `blockify` was not included because it has a variadic parameter.

### CODE REVIEW

File: src/wasm-builder.h [hunk 1]
```diff
@@ -1336,7 +1285,9 @@ class Builder {
 
   // ensure a node is a block, if it isn't already, and optionally append to the
   // block
-  Block* blockify(Expression* any, Expression* append = nullptr) {
+  Block* blockify(Expression* any,
+                  Expression* append = nullptr,
+                  std::optional<Type> type = std::nullopt) {
```
Comment (tlively) [inline src/wasm-builder.h:1290]: Oh neat, does this mean we can pass a type to the variadic version of `blockify` below now, too?

File: src/wasm/wasm.cpp [hunk 2]
```diff
@@ -174,85 +172,82 @@ handleUnreachable(Block* block,
...
-    // Nothing branches here, so this is easy.
-    handleUnreachable(this, NoBreak);
-    return;
-  }
+void Block::finalize(std::optional<Type> type_, Breakability breakability) {
+  if (type_) {
+    type = *type_;
+    if (type == Type::none && list.size() > 0) {
+      handleUnreachable(this, breakability);
+    }
```
Comment (kripken) [inline src/wasm/wasm.cpp:180]: Maybe early-return after this to avoid indenting all the rest in an else?

**Review by kripken — APPROVED** [review]
Nice! I had only thought to do this in wasm-builder so very good idea to also do it in wasm.h as well.

lgtm with early returns + also please do a quick check to see this does not regress speed.

File: src/wasm/wasm.cpp [hunk 3]
```diff
@@ -174,85 +172,82 @@ handleUnreachable(Block* block,
...
-    handleUnreachable(this, breakability);
+    // The default type is according to the value that flows out.
+    BranchUtils::BranchSeeker seeker(this->name);
+    Expression* temp = this;
+    seeker.walk(temp);
+    if (seeker.found) {
+      // Calculate the supertype of the branch types and the flowed-out type. If
+      // there is no supertype among the available types, assume the current
+      // type is already correct. TODO: calculate proper LUBs to compute a new
+      // correct type in this situation.
```
Comment (tlively) [inline src/wasm/wasm.cpp:204]: This comment looks extremely stale since we do calculate LUBs now.

File: src/wasm/wasm.cpp [hunk 4]
```diff
@@ -874,25 +869,26 @@ void TableCopy::finalize() {
...
-  type = type_;
-  bool allUnreachable = body->type == Type::unreachable;
-  for (auto catchBody : catchBodies) {
-    allUnreachable &= catchBody->type == Type::unreachable;
-  }
-  if (type == Type::none && allUnreachable) {
-    type = Type::unreachable;
+  } else {
+    // If none of the component bodies' type is a supertype of the others,
+    // assume the current type is already correct. TODO: Calculate a proper LUB.
```
Comment (tlively) [inline src/wasm/wasm.cpp:885]: Also stale.

**Review by tlively — APPROVED** [review]

**Comment by aheejin** [comment]
Huh, gcc doesn't seem to like what I have here...

File: src/wasm/wasm.cpp (hunk 2 above)
Comment (aheejin) [reply src/wasm/wasm.cpp:180]: Done. Also did the same for `If::finalize`. (Others (`Loop`, `Try`, and `TryTable`) were kinda small and didn't seem worth it)

File: src/wasm/wasm.cpp (hunk 3 above)
Comment (aheejin) [reply src/wasm/wasm.cpp:204]: Fixed.

File: src/wasm/wasm.cpp (hunk 4 above)
Comment (aheejin) [reply src/wasm/wasm.cpp:885]: Fixed

File: src/wasm-builder.h (hunk 1 above)
Comment (aheejin) [reply src/wasm-builder.h:1290]: I thought so, but it doesn't seem to work... It continues to `blockify(any, append)` and errors out as soon as the `append` is not `Expression*`, like this:
```console
/usr/local/google/home/aheejin/binaryen/src/wasm-builder.h:1307:12: error: no matc
hing member function for call to 'blockify'
 1307 |     return blockify(blockify(any, append), args...);
      |            ^~~~~~~~
/usr/local/google/home/aheejin/binaryen/src/passes/StackCheck.cpp:112:25: note: in
 instantiation of function template specialization 'wasm::Builder::blockify<wasm::
Type>' requested here
  112 |     auto *ret = builder.blockify(check, newSet, Type(Type::i32));
      |                         ^
/usr/local/google/home/aheejin/binaryen/src/wasm-builder.h:1288:10: note: candidat
e function not viable: no known conversion from 'wasm::Type' to 'Expression *' for
 2nd argument
 1288 |   Block* blockify(Expression* any,
      |          ^
 1289 |                   Expression* append = nullptr,
      |                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/google/home/aheejin/binaryen/src/wasm-builder.h:1306:10: note: candidat
e function template not viable: no known conversion from 'wasm::Type' to 'Expressi
on *' for 2nd argument
 1306 |   Block* blockify(Expression* any, Expression* append, Ts... args) {
      |          ^                         ~~~~~~~~~~~~~~~~~~
1 error generated.
```

I guess I should revert the `blockify` part. `blockifyWithName` should be fine given that it doesn't use a variadic parameter though.

**Review by tlively — APPROVED** [review]

**Comment by aheejin** [comment]
@kripken 
> also please do a quick check to see this does not regress speed.

I ran `check.py` (% `wasm-reduce` tests, which takes too long time) on `main` and this branch on my local machine, and the time it took (in 3 averaged runs) were 191.44s (main) vs 191.11s (this), which doesn't seem to be a meaningful difference.

**Pull request merged**